// @pnp/sp/files/readable-file.js
import { BlobParse, BufferParse, JSONParse, TextParse } from "@pnp/queryable/index.js";
import { _SPInstance, SPQueryable } from "../spqueryable.js";
export class ReadableFile extends _SPInstance {
    /**
     * Gets the contents of the file as text. Not supported in batching.
     *
     */
    getText() {
        return this.getParsed(TextParse());
    }
    /**
     * Gets the contents of the file as a blob, does not work in Node.js. Not supported in batching.
     *
     */
    getBlob() {
        return this.getParsed(BlobParse());
    }
    /**
     * Gets the contents of a file as an ArrayBuffer, works in Node.js. Not supported in batching.
     */
    getBuffer() {
        return this.getParsed(BufferParse());
    }
    // (headers({ "binaryStringResponseBody": "true" })
    /**
     * Gets the contents of a file as an ArrayBuffer, works in Node.js. Not supported in batching.
     */
    getJSON() {
        return this.getParsed(JSONParse());
    }
    getParsed(parser) {
        return SPQueryable(this, "$value").using(parser)();
    }
}
